### PR TITLE
AB#38602 error message switch modes

### PR DESCRIPTION
**Pull request by lhao03**
## Overview

If you want to try out a query but you're in the wrong mode, this error will tell you to switch to another mode. 

### Demo

image
error-message
error

### Notes

The branch I based off of is when interns/feature/switch-permissions still had the cat button.
The Status messages are usually being generated based on the response from query. For this PR, the Status messages are being generated based on the permissions. 

## Testing Instructions

* pull the branch
* run the code

### CODE REVIEW

File: src/app/services/reducers/permission-mode-reducer.ts [hunk 1]
```diff
@@ -15,3 +15,4 @@ export function permissionModeType(state: boolean = DISPLAY_DELEGATED_PERMISSION
             return state;
     }
 }
+
```
Comment (gracet9182) [inline src/app/services/reducers/permission-mode-reducer.ts]: Extra line

File: src/app/views/App.tsx [hunk 2]
```diff
@@ -86,6 +88,7 @@ class App extends Component<IAppProps, IAppState> {
   public componentDidMount = async () => {
     this.displayToggleButton(this.mediaQueryList);
     this.mediaQueryList.addListener(this.displayToggleButton);
+    this.props.actions!.fetchScopes();
```
Comment (gracet9182) [inline src/app/views/App.tsx:91]: Is this necessary? I removed it and it still worked

File: src/app/views/app-sections/StatusMessages.tsx [hunk 3]
```diff
@@ -51,16 +52,31 @@ export function statusMessages(queryState: any, sampleQuery: IQuery, actions: an
...
+      return permissionModeType === DISPLAY_APPLICATION_PERMISSIONS ? permissions.filter(isRSC) : permissions;
+    }
+
+    const permissions: IPermission[] = scopes.hasUrl ? scopes.data : [];
+    const filteredScopes = filterPermissionsForRSC();
+    const noPerms = filteredScopes.length === 0;
 
     return (
-      <MessageBar messageBarType={messageType}
+      <MessageBar messageBarType={noPerms ? 5 : messageType}
```
Comment (gracet9182) [inline src/app/views/app-sections/StatusMessages.tsx]: Is there a particular reason it's hard coded with a 5?

**Review by gracet9182 — COMMENTED** [review]
Great job :)!

File: src/app/services/reducers/permission-mode-reducer.ts (hunk 1 above)
Comment (lhao03) [reply src/app/services/reducers/permission-mode-reducer.ts]: yeet

File: src/app/views/app-sections/StatusMessages.tsx (hunk 3 above)
Comment (lhao03) [reply src/app/views/app-sections/StatusMessages.tsx]: for `messageBarType`, 5 shows the yellow card: 
image

but i should of made it a constant

**Review by gracet9182 — APPROVED** [review]

**Review by ayubun — CHANGES_REQUESTED** [review]
There are a couple bugs I have noticed while just tinkering around, but I am not quite sure where in the code they might be yet.

For one, some items (specifically, those with tips in their sample query) tend to have this message, and then have it flicker away once the tip arrives. Maybe a solution to this would be using a red error as opposed to the yellow tip?

https://user-images.githubusercontent.com/49354780/125085370-5c5d0300-e07f-11eb-9e22-cb16a34165ce.mp4

The second bug I noticed was this weird issue that said I needed to be delegated user as an app, and also said I needed to be an app as a delegated user. I haven't reproduced it yet but I also haven't given it much time.

https://user-images.githubusercontent.com/49354780/125086387-6f240780-e080-11eb-944b-04dcabaa052c.mp4

Both of these bugs were noticeable within the first 3 minutes or so of tinkering with the code. Because of that, I am unsure if this PR is ready for merge until they get fixed-- unless we feel like we need the content merged and can fix the bug later if we just create an Issue for it.

**Comment by lhao03** [comment]
> There are a couple bugs I have noticed while just tinkering around, but I am not quite sure where in the code they might be yet.
> 
> For one, some items (specifically, those with tips in their sample query) tend to have this message, and then have it flicker away once the tip arrives. Maybe a solution to this would be using a red error as opposed to the yellow tip?
> 
>  Graph.Explorer.-.Google.Chrome.2021-07-09.06-27-50.mp4 
> The second bug I noticed was this weird issue that said I needed to be delegated user as an app, and also said I needed to be an app as a delegated user. I haven't reproduced it yet but I also haven't given it much time.
> 
>  Graph.Explorer.-.Google.Chrome.2021-07-09.06-26-09.online-video-cutter.com.mp4 
> Both of these bugs were noticeable within the first 3 minutes or so of tinkering with the code. Because of that, I am unsure if this PR is ready for merge until they get fixed-- unless we feel like we need the content merged and can fix the bug later if we just create an Issue for it.

Thanks @AyuAi for noticing this buggy behavior! I agree, it's not ready to be merged.

The reason for the first bug is because the permissions error is kinda fighting with the tip and I'm not grabbing the scopes fast enough. 
image

Basically for some instance in time, there seems to be no scopes -> triggering the message about switching modes, but when the scopes are finally fetched or there is a tip, it flickers away. I'll have to work on smoothing that out.

For the second bug, when there are no application permissions it tells the user to switch to user mode, and vice versa. However it seems that sometimes there's no permissions for both user mode and app mode for this query you tried `https://graph.microsoft.com/v1.0/groups/group-id-for-teams/drive/root/children` so we might have to figure out how to deal with this edge case.

**Comment by lhao03** [comment]
Closing because the solution is very hacky and waiting on the endpoint to be added to DevX API before redoing the solution to this ticket.

**Pull request closed without merging**